### PR TITLE
Use a union to eliminate an aliasing error

### PR DESCRIPTION
**Pull request by rwy0717**
Introduce a new type, DoubleConversionData, for converting between a
double and it's bits (uint64_t). Use this type to CAS the bits of a
double. Eliminates a bad cast from double* to uint64_t*.

Fixes: #3123
Signed-off-by: Robert Young <rwy0717@gmail.com>

### CODE REVIEW

**Comment by rwy0717** [comment]
I still need to double check that this eliminates the warning (without adding new ones) everywhere. Marking as WIP until I've tested on more configurations.

**Comment by Leonardo2718** [comment]
Unfortunately, I think this is still relying on undefined behaviour (whether or not the compiler emits a warning for it or not). For reference, see the gcc documentation for `-fstrict-aliasing` https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html.

**Comment by rwy0717** [comment]
Yes, this is still undefined behaviour. It would be nice to have a "true" CAS for doubles, but it looks like there is very little compiler support for it. I think we're stuck with casting/UB for the foreseeable future, so this was my middleground solution.

Available:
- newer gcc and any clang have a generic `__atomic_compare_exchange`
- zos xlc has `__csg` and maybe `cds` which operate on `void*`

Not Available:
- older gcc, only have `__sync` intrinsics
- aix xlc and msvc intrinsics only operate on `volatile long`

**Comment by rwy0717** [comment]
I don't think memcpy can be used to implement an atomic cas.

**Comment by Leonardo2718** [comment]
Yeah, ok. Now that I've thought about it, you can only use `memcpy` to fix the undefined behaviour with `oldValue` and `newValue`, but not `data`. In this case, I suggest clear documentation describing the potential issue and just going with what you have. :slightly_smiling_face:

**Comment by rwy0717** [comment]
I can look at implementing a `lockCompareExchangeDouble`, which does the right thing if the generic intrinsics are available, and only fall back to "integral cas" if needed.

**Comment by rwy0717** [comment]
After looking it at it a bit more, I realized that the newer `__atomic_compare_exchange` (clang+gcc) returns a boolean, rather than the value compared. This means we can't implement the existing API using the newer `__atomic*` intrinsics.

**Comment by Leonardo2718** [comment]
> After looking it at it a bit more, I realized that the newer __atomic_compare_exchange (clang+gcc) returns a boolean, rather than the value compared. This means we can't implement the existing API using the newer __atomic* intrinsics.

That's disappointing. :disappointed:

**Comment by rwy0717** [comment]
OK It appears this PR eliminates the warning on gcc 7. Removing the WIP.

**Comment by Leonardo2718** [comment]
@genie-omr build all

File: include_core/AtomicSupport.hpp [hunk 1]
```diff
@@ -643,16 +648,19 @@ class VM_AtomicSupport
 	{
 		/* double is stored as 64bit */
 		/* Stop compiler optimizing away load of oldValue */
-		volatile uint64_t *localAddr = (volatile uint64_t *)address;
+		volatile DoubleConversionData* data = (volatile DoubleConversionData*)address;
```
[the remark this answers is not in the document]
Comment (rwy0717) [reply include_core/AtomicSupport.hpp]: Updated with a comment.

**Comment by charliegracie** [comment]
@genie-omr build all

**Pull request merged**